File: user-service/src/main/java/rs/raf/user_service/specification/ClientSearchSpecification.java
package rs.raf.user_service.specification;

import org.springframework.data.jpa.domain.Specification;
import rs.raf.user_service.domain.entity.Client;
import rs.raf.user_service.domain.entity.Employee;

public class ClientSearchSpecification {

    //Pretraga po imenu
    public static Specification<Client> firstNameContains(String firstName) {
        return (root, query, criteriaBuilder) -> {
            if (firstName == null || firstName.isEmpty()) {
                return criteriaBuilder.conjunction();
            }
            return criteriaBuilder.like(
                    criteriaBuilder.lower(root.get("firstName")),
                    "%" + firstName.toLowerCase() + "%"
            );
        };
    }

    //Pretraga po prezimenu
    public static Specification<Client> lastNameContains(String lastName) {
        return (root, query, criteriaBuilder) -> {
            if (lastName == null || lastName.isEmpty()) {
                return criteriaBuilder.conjunction();
            }
            return criteriaBuilder.like(
                    criteriaBuilder.lower(root.get("lastName")),
                    "%" + lastName.toLowerCase() + "%"
            );
        };
    }

    //Pretrega po mailu
    public static Specification<Client> emailContains(String email) {
        return (root, query, criteriaBuilder) -> {
            if (email == null || email.isEmpty()) {
                return criteriaBuilder.conjunction();
            }
            return criteriaBuilder.like(
                    criteriaBuilder.lower(root.get("email")),
                    "%" + email.toLowerCase() + "%"
            );
        };
    }

    public static Specification<Client> hasRole(String roleName) {
        return (root, query, cb) -> roleName == null ? null :
                cb.equal(cb.lower(root.join("role").get("name")), roleName.toLowerCase());
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/specification/ClientSearchSpecification.kt
package pack.userservicekotlin.specification

import jakarta.persistence.criteria.CriteriaBuilder
import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Root
import org.springframework.data.jpa.domain.Specification
import pack.userservicekotlin.domain.entities.Client
import java.util.*

object ClientSearchSpecification {
    // Pretraga po imenu
    fun firstNameContains(firstName: String?): Specification<Client?> {
        return Specification<Client?> { root: Root<Client?>, query: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
            if (firstName == null || firstName.isEmpty()) {
                return@Specification criteriaBuilder.conjunction()
            }
            criteriaBuilder.like(
                criteriaBuilder.lower(root.get<String>("firstName")),
                "%" + firstName.lowercase(Locale.getDefault()) + "%",
            )
        }
    }

    // Pretraga po prezimenu
    fun lastNameContains(lastName: String?): Specification<Client?> {
        return Specification<Client?> { root: Root<Client?>, query: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
            if (lastName == null || lastName.isEmpty()) {
                return@Specification criteriaBuilder.conjunction()
            }
            criteriaBuilder.like(
                criteriaBuilder.lower(root.get<String>("lastName")),
                "%" + lastName.lowercase(Locale.getDefault()) + "%",
            )
        }
    }

    // Pretrega po mailu
    fun emailContains(email: String?): Specification<Client?> {
        return Specification<Client?> { root: Root<Client?>, query: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
            if (email == null || email.isEmpty()) {
                return@Specification criteriaBuilder.conjunction()
            }
            criteriaBuilder.like(
                criteriaBuilder.lower(root.get<String>("email")),
                "%" + email.lowercase(Locale.getDefault()) + "%",
            )
        }
    }
}
